hail/python/hail/methods/statgen.py

```diff
@@ -446,6 +447,146 @@ def linear_regression_rows(y, x, covariates, block_size=16, pass_through=()) ->
+
+    # Given a hail array, get the mean of the nonmissing entries and
+    # return new array where the missing entries are the mean.
+    def mean_impute(hl_array):
+        non_missing_mean = hl.mean(hl_array, filter_missing=True)
+        return hl.map(lambda arr_entry: hl.if_else(hl.is_defined(arr_entry), arr_entry, non_missing_mean), hl_array)
+
+    def select_array_indices(hl_array, indices):
+        return indices.map(lambda i: hl_array[i])
+
+    def dot_rows_with_themselves(matrix):
+        return (matrix * matrix) @ hl.nd.ones(matrix.shape[1])
+
+    ht_local = mt._localize_entries(entries_field_name, sample_field_name)
+
+    ht = ht_local.transmute(**{entries_field_name: ht_local[entries_field_name][x_field_name]})
+    ht = ht._group_within_partitions("grouped_fields", block_size)
+
+    ys_and_covs_to_keep_with_indices = hl.zip_with_index(ht[sample_field_name]).filter(lambda struct_with_index: all_defined(struct_with_index[1], y_field_names + cov_field_names))
+    indices_to_keep = ys_and_covs_to_keep_with_indices.map(lambda pair: pair[0])
+    ys_and_covs_to_keep = ys_and_covs_to_keep_with_indices.map(lambda pair: pair[1])
+
+    cov_nd = hl.nd.array(ys_and_covs_to_keep.map(lambda struct: hl.array([struct[cov_name] for cov_name in cov_field_names]))) if cov_field_names else hl.nd.zeros((hl.len(indices_to_keep), 0))
+    ht = ht.annotate_globals(kept_samples=indices_to_keep,
+                             __y_nd=hl.nd.array(ys_and_covs_to_keep.map(lambda struct: hl.array([struct[y_name] for y_name in y_field_names]))),
+                             __cov_nd=cov_nd)
+    k = builtins.len(covariates)
+
+    ht = ht.annotate(**{X_field_name: hl.nd.array(hl.map(lambda row: mean_impute(select_array_indices(row, ht.kept_samples)), ht["grouped_fields"][entries_field_name])).T})
+    n = hl.len(ht.kept_samples)
+    ht = ht.annotate_globals(d=n-k-1)
+    cov_Qt = hl.if_else(k > 0, hl.nd.qr(ht.__cov_nd)[0].T, hl.nd.zeros((0, n)))
+    ht = ht.annotate_globals(__Qty=cov_Qt @ ht.__y_nd)
+    ht = ht.annotate_globals(__yyp=dot_rows_with_themselves(ht.__y_nd.T) - dot_rows_with_themselves(ht.__Qty.T))
+
+    sum_x_nd = ht[X_field_name].T @ hl.nd.ones((n,))
+    ht = ht.annotate(sum_x=nd_to_array(sum_x_nd))
+    ht = ht.annotate(__Qtx=cov_Qt @ ht[X_field_name])
+    ht = ht.annotate(__ytx=ht.__y_nd.T @ ht[X_field_name])
+    ht = ht.annotate(__xyp=ht.__ytx - (ht.__Qty.T @ ht.__Qtx))
+    ht = ht.annotate(__xxpRec=(dot_rows_with_themselves(ht[X_field_name].T) - dot_rows_with_themselves(ht.__Qtx.T)).map(lambda entry: 1 / entry))
+    ht = ht.annotate(__b=ht.__xyp * ht.__xxpRec)
+    ht = ht.annotate(__se=((1.0/ht.d) * (ht.__yyp.reshape((-1, 1)) @ ht.__xxpRec.reshape((1, -1)) - (ht.__b * ht.__b))).map(lambda entry: hl.sqrt(entry)))
+    ht = ht.annotate(__t=ht.__b / ht.__se)
+    ht = ht.annotate(__p=ht.__t.map(lambda entry: 2 * hl.expr.functions.pT(-hl.abs(entry), ht.d, True, False)))
```

Comment:
You don't have to do this now, but I think this could be written as a single annotate, with only those values that are used later defined as fields, and with shared subcomputations lifted out to python variables. E.g.
```
Qtx = cov_Qt @ ht[X_field_name]
ytx = ht.__y_nd.T @ ht[X_field_name]
xyp = ht.__ytx - (__Qty.T @ ht.__Qtx)
b = xyp * xxpRec)
...
ht = ht.annotate(__ytx=ytx, ...)
```